### PR TITLE
Draft for manual paging implementation.

### PR DESCRIPTION
**Pull request by Kircheneer**
Ref: https://github.com/nautobot/pynautobot/issues/106

### CODE REVIEW

**Comment by Kircheneer** [comment]
Should look at pushing the entire logic into `Endpoint`, allowing `limit` and `offset` in kwargs instead. Keeps the API for the end user the same but should be cleaner internally. Also removes the special treatment for `limit` and `offset` as much as possible, as in the end they are just normal REST API parameters.

File: pynautobot/core/endpoint.py [hunk 1]
```diff
@@ -232,13 +254,19 @@ def filter(self, *args, api_version=None, **kwargs):
             raise ValueError("A reserved {} kwarg was passed. Please remove it " "try again.".format(RESERVED_KWARGS))
 
         api_version = api_version or self.api.api_version
+
+        manual_pagination = self._check_manual_pagination(limit, offset)
+        url = "{}/".format(self.url)
+        if manual_pagination:
+            url += "?" + urllib.parse.urlencode({"limit": limit, "offset": offset})
```
Comment (jvanderaa) [inline pynautobot/core/endpoint.py:261]: If this sticks around, let's look at passing it to the Requests library with `params`.

File: pynautobot/core/query.py [hunk 2]
```diff
@@ -303,6 +305,8 @@ def req_all():
             req = self._make_call(add_params=add_params)
             if isinstance(req, dict) and req.get("results") is not None:
                 ret = req["results"]
+                if self.manual_pagination:
+                    return ret
```
Comment (jvanderaa) [inline pynautobot/core/query.py:309]: Add some commentary on this, I'm not clear why returning here.

File: pynautobot/core/endpoint.py [hunk 3]
```diff
@@ -80,7 +82,22 @@ def _lookup_ret_obj(self, name, model):
...
+    def _check_manual_pagination(self, limit, offset):
+        """Checks whether manual pagination should be applied.
+
+        :arg limit: Integer value corresponding to the API parameter.
+        :arg offset: Integer value corresponding to the API parameter.
+
+        :Returns: True or false, depending on whether manual pagination is used."""
+        if limit:
+            if not offset:
+                raise ValueError("If limit is passed to the 'all' method, offset needs to be passed as well.")
```
Comment (Kircheneer) [inline pynautobot/core/endpoint.py:94]: This should be `if offset is None`, otherwise it doesn't allow 0.

**Comment by joewesch** [comment]
This PR is really stale and it understandably now has conflicts. Closing to clean up, but feel free to reopen if you want to pick it back up again in the future.

**Pull request closed without merging**